### PR TITLE
Wrong statement regarding weight decay in pytorch

### PR DESCRIPTION
**Pull request by finale80**
The two sentences 
*[...] By default, PyTorch decays both weights and biases simultaneously. Here, we only set weight_decay for the weight, so the bias parameter b will not decay [...]*
contradict each other. More specifically, I believe the 2nd one is wrong.

As from PyTorch implementation https://github.com/pytorch/pytorch/blob/acd6ca8cfa9537284928fb5d36834d1e5ae1e6f3/torch/optim/sgd.py#L120
all parameters receive the same weight decay.
So the first sentence is correct.

In the example of this chapter, to avoid decaying the bias, one would need to use the following
```
...
   def loss(self, y_hat, y):
        return super().loss(y_hat, y) + self.weight_decay * 0.5 * (self.net.weight**2).sum()

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=self.lr, weight_decay=0)
...
```

I attach here a self-contained example based on this chapter
(note: I used .txt to bypass file extension checking).

The code compares 3 versions of the model
1. the default implementation (i.e., the one suggested by the book which decays all parameters)
2. a "manual" implementation of the default behaviour (to show that the default is indeed decaying all parameters)
3. the option mentioned above (which does not decay the bias).

Overall, the simplest way to fix the problem is to drop the 2nd sentence as proposed.

Alternatively, one could add the comparison as from the code attached. But
1. I need to be extended to Tensorflow and MXNet
6. I think the right way to skip the decay would be by specifying the parameter differently (I believe one could group parameters with different options). But this would be too much details given the aim of the chapter

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

[chapter3.7_weight-decay.py.txt](https://github.com/d2l-ai/d2l-en/files/9454152/chapter3.7_weight-decay.py.txt)

### CODE REVIEW

**Review by AnirudhDagar — COMMENTED** [review]
Thanks @finale80 for your PR!

This is a good catch and makes sense now. I remember adding this statement earlier because it was actually relevant at the time of writing the code for pytorch version (see #912) in the old d2l api. We used to only decay the weight params and not the bias.

So we had something like the following to handle weight decay separately for pytorch:

```python
      return torch.optim.SGD([
          {"params":self.net.weight,'weight_decay': self.wd},
          {"params":self.net.bias}], lr=self.lr)
```

I see that the above code was refactored and changed in #2034 and the text related to the code change was probably missed.

@cheungdaven after PR #2034 there seems to be an inconsistency between pytorch and mxnet. Specifically mxnet doesn't decay bias but pytorch does. I'm curious if there was any specific reason to remove [these L394-396](https://github.com/d2l-ai/d2l-en/pull/2034/files#diff-8be31f544cfaa6008c11c863dba743c9a4abb4a0436abb2763810b701d41fed0L394-L396)?

**Comment by cheungdaven** [comment]
It makes sense to me. feel free to change it back @AnirudhDagar

**Comment by AnirudhDagar** [comment]
Thanks @cheungdaven! @finale80 could you please add the params grouping back in pytorch code like we had before instead of editing the text?

**Comment by finale80** [comment]
@AnirudhDagar done!

More specifically
* I've updated the code
* I re-integrated the original statement in the text (removed in my PR) and added minor modification to the text (to make it even more clear).

I also double checked the code and indeed everything is consistent now.
I attach my updated script.](url)
[chapter3.7_weight-decay.py.txt](https://github.com/d2l-ai/d2l-en/files/9540129/chapter3.7_weight-decay.py.txt)

File: chapter_linear-regression/weight-decay.md [hunk 1]
```diff
@@ -392,10 +395,16 @@ class WeightDecay(d2l.LinearRegression):
...
-        return torch.optim.SGD(self.net.parameters(), 
-                               lr=self.lr, weight_decay=self.wd)
+        """Defined in :numref:`sec_linear_concise`"""
+        return torch.optim.SGD(
+            [
+                {"params": self.net.weight, "weight_decay": self.wd},
+                {"params": self.net.bias},
+            ],
+            lr=self.lr,
+        )
```
Comment (AnirudhDagar) [inline chapter_linear-regression/weight-decay.md:407]:
```suggestion
        return torch.optim.SGD([
            {'params': self.net.weight, 'weight_decay': self.wd},
            {'params': self.net.bias}], lr=self.lr)
```

**Review by AnirudhDagar — CHANGES_REQUESTED** [review]
@finale80 just a minor style comment below (the idea is to minimize usage of lines wherever possible), the rest looks good. Thanks, will merge this after the change.

File: chapter_linear-regression/weight-decay.md [hunk 2]
```diff
@@ -392,10 +395,12 @@ class WeightDecay(d2l.LinearRegression):
         super().__init__(lr)
         self.save_hyperparameters()
         self.wd = wd
-    
+
     def configure_optimizers(self):
-        return torch.optim.SGD(self.net.parameters(), 
-                               lr=self.lr, weight_decay=self.wd)
+        """Defined in :numref:`sec_linear_concise`"""
```
Comment (AnirudhDagar) [inline chapter_linear-regression/weight-decay.md:400]: Also no need to add this line.
```suggestion
```

**Review by AnirudhDagar — APPROVED** [review]
